### ISSUE TITLE
Entering DFU doesn't always work

### ISSUE THREAD

**Issue by jkasten2** [underverk/SmartWatch#4]
I am able to get into DFU mode when I need to but it can take me up to 10 minutes to get it there sometimes to test a code change!

The timing required to enter DFU mode to picky. If you insert the USB cable to soon it will lock the watch and require to hold the power button for 15 seconds minimum. Is there anther way to reliably enter DFU mode?

Maybe a way to programmaticly enter DFU mode with a line of code? This would speed up testing code changes greatly for me!

Thanks.

**Comment by dcuartielles** [comment]
As far as we know it should be possible to make a piece of code in ARM Assembly that could be calling the DFU bootloader, but it would also require some mechanism to trigger that mode.

That said, we are not planning on doing that any time soon, but we are happy to get in a pull-request including such funcionality. Are you up for the task?

**Comment by jkasten2** [comment]
I am up for the task but I need some help.

Good to know it is possible to trigger the mode programmaticly. I have been trying to get this to work from 2 different angles.

#(1) Call the bootloader entry point where the DFU mode is started. I tried 0x1FFF7800 but it just locks the watch. (Do I need to do something else first like disable interrupts or clear something more?)

#(2) Look into turn the mode on while the watch is running my current firmware. I tried running the following 2 lines but it didn't seem to do anything.(The firmware loop keeps going like normal)

RCC_AHB2PeriphClockCmd(RCC_AHB2Periph_OTG_FS, ENABLE);
NVIC_EnableIRQ(OTG_FS_IRQn);

I am thinking #(1) would be easier as I would just be calling pre existing code on the watch but maybe it isn't that simple.

With #(2) I am thinking I might need to do some more setup like setting some GPIO pins or maybe I need to use OTG_FS_WKUP_IRQn instead?

Is their a direction that is better to go down, something I am overlooking, or links to documents I should read would be appreciated!

**Comment by jkasten2** [comment]
I have been making a some progress. Here is some code that gets the USB to detect on my pc but DFU mode isn't working though or it is in some different mode:

void StartBootloaderDFUMode() {
  // 0x1FFF77E2 // just frezzes this one can't be right
  // 0x1fff0004 // USB device detects on PC!!! W00t! Only it doesn't seem to work.
                // Windows show it as an unknown device and after forcing the driver i just get this
                // "A request for the USB device descriptor failed."

  __set_PRIMASK(1); // Disable interrupts

  SysMemBootJump=(void (_)(void)) (_((u32 *) 0x1fff0004));

// A few functions commented out as having them all disable just gets me a black screen with no USB showing up.

  //The following lines is what USART_DeInit(); does
    RCC_APB2PeriphResetCmd(RCC_APB2Periph_USART1, ENABLE);
    RCC_APB2PeriphResetCmd(RCC_APB2Periph_USART1, DISABLE);
    RCC_APB1PeriphResetCmd(RCC_APB1Periph_USART2, ENABLE);
    RCC_APB1PeriphResetCmd(RCC_APB1Periph_USART2, DISABLE);
    RCC_APB1PeriphResetCmd(RCC_APB1Periph_USART3, ENABLE);
    RCC_APB1PeriphResetCmd(RCC_APB1Periph_USART3, DISABLE);
    RCC_APB1PeriphResetCmd(RCC_APB1Periph_UART4, ENABLE);
    RCC_APB1PeriphResetCmd(RCC_APB1Periph_UART4, DISABLE);
    RCC_APB1PeriphResetCmd(RCC_APB1Periph_UART5, ENABLE);
    RCC_APB1PeriphResetCmd(RCC_APB1Periph_UART5, DISABLE);
    RCC_APB2PeriphResetCmd(RCC_APB2Periph_USART6, ENABLE);
    RCC_APB2PeriphResetCmd(RCC_APB2Periph_USART6, DISABLE);

  RCC_DeInit();
  // Next 2 lines are what SYSCFG_DeInit(); does
   RCC_APB2PeriphResetCmd(RCC_APB2Periph_SYSCFG, ENABLE);
   RCC_APB2PeriphResetCmd(RCC_APB2Periph_SYSCFG, DISABLE);
  //PWR_DeInit();
  // Next 2 lines are what I2C_DeInit(I2C1); does
    RCC_APB1PeriphResetCmd(RCC_APB1Periph_I2C1, ENABLE);
    RCC_APB1PeriphResetCmd(RCC_APB1Periph_I2C1, DISABLE);
  SysTick->CTRL = 0x04;  // reset Systick Timer // Tried 0x00 too both seem to be ok
  SysTick->LOAD = 0;
  SysTick->VAL = 0;
  RCC_SYSCLKConfig(RCC_SYSCLKSource_HSI); // Select HSI as system clock source

  /*
  GPIO_DeInit(GPIOA);
  GPIO_DeInit(GPIOB);
  GPIO_DeInit(GPIOC);
  GPIO_DeInit(GPIOD);
  GPIO_DeInit(GPIOE);
  GPIO_DeInit(GPIOF);
  GPIO_DeInit(GPIOG);
  GPIO_DeInit(GPIOH);
  GPIO_DeInit(GPIOI);
  */

  //__set_MSP(*(volatile unsigned int *)0x1FFFF000); // this locks and no USB detected.
  __set_MSP(0x20001000); // Set the main stack pointer to its default value

  SysMemBootJump();
  while(1);
}

I call this from my firmware by looking for a touch in the corner of the screen. It locks the watch screen and on my pc in windows it detects an unknown USB device. When I force install the smartwatch driver I just says "A request for the USB device descriptor failed." with a start code 10. It only does this though if I trigger this function when the USB cable is disconnected and then connect it to my pc. Do I need to just reset something more for it to work? Or do i need to set the pins to USB before I call the bootloader?

Anther thing I was trying to do is trigger the bootloader at the boot of my firmware. In my firmware I run the following 2 lines:
*((unsigned long *)0x2000FFF0) = 0xDEADBEEF;          // Set a trigger value at mem address 0x2000FFF0
SCB->AIRCR = AIRCR_VECTKEY_MASK | (u32)0x04;  // Reboot the watch

I added the following to the top of Reset_Handler in startup_stm32f2xx.S.
  LDR        R0, =0x2000FFF0
  LDR        R1, =0xDEADBEEF
  LDR        R2, [R0, #0]
  STR        R0, [R0, #0]
  CMP        R2, R1
  BEQ        Reboot_Loader

.... then this after   .type  Reset_Handler, %function

Reboot_Loader:
    LDR       R0, =0x1FFFF000
    LDR       SP,[R0, #0]
    LDR       R0,[R0, #4]
    BX        R0

However it it just seems to look-up. Just to make sure the bootloader was getting called I move the 4 lines of ASM from Reboot_loader at the top of Reset_Handler and same thing. Just locks the watch with nothing on the screen and my PC doesn't see a USB device. Do you think it is maybe just in a USART or CAN2 mode when I do this instead of DFU?

Also dumb question but where does the green dot and line get drawn from? Does that get done somewhere in 1FFFF000? I am wondering if I should be seeing this when calling the bootloader.

Thanks!

Some of the code above came from the following 2 sources:
https://my.st.com/public/STe2ecommunities/mcu/Lists/cortex_mx_stm32/Threaded.aspx?RootFolder=%2Fpublic%2FSTe2ecommunities%2Fmcu%2FLists%2Fcortex_mx_stm32%2FJump%20to%20internal%20bootloader&FolderCTID=0x01200200770978C69A1141439FE559EB459D7580009C4E14902C3CDE46A77F0FFD06506F5B&currentviews=1653

http://www.youtube.com/watch?v=cvKC-4tCRgw

**Comment by underverk** [comment]
I'm a bit surprised to see the 1FFF0004 address for doing this, isn't this just a shadow of the actual bootloader area?
I believe (but correct me if I'm wrong) that the bootloader is located at 0x08000000. The entry point address should as such be located at 0x08000004.

I remember speaking with someone from Sony about this, and it was his firm belief that there was no (intentional) way of entering the DFU bootloader programatically. This might not be correct, and regardless, everything is possible - but it may take some doing.

I've had issues getting into DFU as well, and it can be frustrating. As you mention, timing is important but when done correctly, getting back into DFU should never take more than 20-25 seconds. I've noticed that you have to wait for a few seconds after 15sek-killing the watch before doing the enter DFU process, or it will immediately attempt to launch the faulty firmware again. For entering DFU it is important that you press and hold the button shortly _before_ plugging it in, and not doing it simultaneously - which fails quite often.

**Comment by jkasten2** [comment]
I found 0x1FFF0000 on page 56 in http://www.st.com/st-web-ui/static/active/en/resource/technical/document/application_note/CD00167594.pdf so I believe that is correct. I tried 0x08000004 as you suggested from my firmware(triggered by touch location) and it just reboots. It states that 0x1FFF0000 is "System memory"(Maybe this is read only then?) and 0x08000000 is the start of the flash.

Someone from Sony didn't think it was possible either, then this will probably be tricky. I wounder if am even close with the function above getting the watch to at lest to show up as a USB device on my pc. Anyway I wonder if it would be easier to put a DFU implantation into my own firmware. I found these this project that has does this github.com/gsmcmullin/blackmagic for an stm32 device. And found this project too using it github.com/paparazzi/luftboot

Thanks for the DFU mode tip, I am able to get into DFU mode much more often now! I was waiting too long before, and your right simultaneously doesn't work either and both lock up the watch even with a working firmware.
